x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -560,6 +561,34 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
     return await this.baseClient.removeReferencesTo(type, id, options);
   }
 
+  public async openPointInTimeForType(
+    type: string | string[],
+    options: SavedObjectsOpenPointInTimeOptions
+  ) {
+    try {
+      const args = { type, options };
+      await this.ensureAuthorized(type, 'open_point_in_time', undefined, { args });
```

Comment:
Which privilege is `open_point_in_time` associated with?

Reply:
The way this is currently written, only superusers will be authorized to open a point-in-time: 
`open_point_in_time` hasn't yet been registered as a saved object action, so our authorization service can't authorize anyone for this, except for superusers who have access to everything:

https://github.com/elastic/kibana/blob/caa5da26f07bdce5dbc8c8f8928d28336af09640/x-pack/plugins/security/server/authorization/privileges/feature_privilege_builder/saved_object.ts#L11-L20

----

> Which privilege is open_point_in_time associated with?

Our two options now are `all` and `read` (see the link to the feature privilege builder above). This one is a bit of a gray area to me, so I need a better understanding of how we plan to use this feature.


How do we expect PITs to be used? Is this something that we expect end users to unknowingly use as a byproduct of using Kibana? If so, what are those scenarios? What happens to the user experience if the user isn't authorized to open a PIT?

Is this instead meant as an administrative function? If so, is it user-initiated, or part of a background job?

Do we know what the expected overhead of opening PITs are, or how frequently we expect them to be opened? I can see from the docs that it will consume extra resources on the cluster (such as heap space), but it's not clear to me what the implications are is it relates to Kibana.

Reply:
> How do we expect PITs to be used? Is this something that we expect end users to unknowingly use as a byproduct of using Kibana? If so, what are those scenarios?

@legrego The motivation for adding PIT support is allowing users to "scroll" through large numbers of saved objects (basically working around the default 10k SO limitation), specifically when doing a SO import/export. This would be something that happens behind the scenes without a user's awareness. I can't really think of a case where an end user would need to be aware that this feature is being used.

> What happens to the user experience if the user isn't authorized to open a PIT?

I guess that's up for discussion, but IMO we should be returning a 403 to the service consuming the API, and they can decide the best way to surface this for end users. 

> Is this instead meant as an administrative function? If so, is it user-initiated, or part of a background job?

Right now it would be a user-initiated administrative function and we'd at a minimum need users who have the ability to import/export to be able to use this. However, as this is meant to be complementary to `SO.find`, I think it would make sense to think of it in terms of users who could perform a search via `SO.find`. Basically you'd be first calling `openPointInTimeForType` with the same types that are subsequently passed to `SO.find`. So if you have permission to call `find` with those types, it stands to reason you should also be able to point the PIT against those same indices.

> Do we know what the expected overhead of opening PITs are, or how frequently we expect them to be opened? I can see from the docs that it will consume extra resources on the cluster (such as heap space), but it's not clear to me what the implications are is it relates to Kibana.

I don't expect these to be opened frequently, as the only real use case is scroll-like functionality over large numbers of saved objects (>10k hits by default). If you have fewer than 10k hits, you can just use the existing `page` / `perPage` functionality. As such, I don't anticipate a significant overhead that would affect Kibana (though I should caveat that while the ES team had indicated there is a degree overhead, we haven't tested what the upper limits would be for Kibana).

> My gut reaction is to classify this as a read operation, but it is technically creating something on the ES side in order to satisfy read operations on the Kibana side.

Yeah I went back and forth on this and could go either way. As you said, we're creating something, but it's only going to be used in the context of reading via `find`.

Reply:
Thanks @lukeelmers that is helpful. If we decide to consider this a `read` operation, **and** if we only expect PITs to be consumed via the `find` API, then is there a benefit to exposing `openPointInTimeForType` as its own public function? Would it make sense to consider PITs a feature of `find`, and adjust `find` to return the PIT-related fields if a PIT was requested/required?

Reply:
> Would it make sense to consider PITs a feature of find, and adjust find to return the PIT-related fields if a PIT was requested/required?

That was one of the ideas [we originally considered](https://github.com/elastic/kibana/issues/86301#issuecomment-767973317), but we ended up leaning toward a more explicit implementation where `find` was doing less magic. That said, we hadn't considered how this could complicate things from an authz perspective, so I'm still open to reverting back to that approach depending on what others think. cc @pgayvallet @rudolf

Reply:
> Right now it would be a user-initiated administrative function and we'd at a minimum need users who have the ability to import/export to be able to use this. However, as this is meant to be complementary to SO.find, I think it would make sense to think of it in terms of users who could perform a search via SO.find

Yea, we need that under the `read` permission group.

> That was one of the ideas we originally considered, but we ended up leaning toward a more explicit implementation where find was doing less magic. That said, we hadn't considered how this could complicate things from an authz perspective, so I'm still open to reverting back to that approach

Given https://github.com/elastic/kibana/pull/89915#discussion_r569232178, and the fact that we already agreed that both approaches were fine and that `openPointInTimeForType` was more of an arbitrary preference between the two, I would be fine reverting back too if we now think that the `openPit` option would be better.

@rudolf?

Reply:
> Do we know what the expected overhead of opening PITs are, or how frequently we expect them to be opened? I can see from the docs that it will consume extra resources on the cluster (such as heap space), but it's not clear to me what the implications are is it relates to Kibana.

I don't know the internal details, but my understanding is that ES needs to do a copy-on-write for every write once a PIT has been opened. So worst case scenario would be opening a PIT and then doing an update by query over all of `.kibana` as this would at least double the disk space consumed by the `.kibana` index. Only once the PIT is closed ES can free up the old copies.

So it feels like we want to restrict this API to an admin-like role. I remember we had discussions to do the same with the import/export API but I can't remember if that was already implemented or something we wanted to do.

Reply:
> I don't know the internal details, but my understanding is that ES needs to do a copy-on-write for every write once a PIT has been opened. So worst case scenario would be opening a PIT and then doing an update by query over all of .kibana as this would at least double the disk space consumed by the .kibana index. Only once the PIT is closed ES can free up the old copies, so opening many PIT's can negatively impact Kibana's stability.
> 
> So it feels like we want to restrict this API to an admin-like role. I remember we had discussions to do the same with the import/export API but I can't remember if that was already implemented or something we wanted to do.


If the primary problem we are solving is the import/export, then we can hide this from the client-side API, and just consume it from within the import/export logic directly. For the time being, it needs to be either a `read` or `all` operation, so technically anyone with access to a SO type will be able to open a PIT if they're authorized for that type, and they have a way to call this function. That's probably OK for now, and I expect that solutions will want to take advantage of this within their own solutions once this merges anyway.

As for the existing import/export APIs: they don't appear to be restricted today -- you can import/export anything that you'd otherwise be able to CRUD. We protect the Saved Objects Management UI to improve the UX, because importing saved objects when you're not authorized for all of them makes for a rather frustrating experience.

Reply:
@legrego @pgayvallet @rudolf Trying to synthesize the various threads, here is where we are at:

> Given https://github.com/elastic/kibana/pull/89915#discussion_r569232178, and the fact that we already agreed that both approaches were fine and that `openPointInTimeForType` was more of an arbitrary preference between the two, I would be fine reverting back too if we now think that the `openPit` option is better.

I spent some time thinking about this and I don't think the implicit API is a good option anymore, because it doesn't deal with the issue of how to close a PIT. (When you are initiating a `find` request, how do you know if you are ready to close a PIT until you get the results back? The only way would be to require you to make an additional unnecessary request. Hence the need for another dedicated `closePointInTime` method).

Which is too bad because I think otherwise the implicit API would solve most of the concerns we've raised WRT security.

> So it feels like we want to restrict this API to an admin-like role.

My only concern with this is that PIT is not really useful unless you are already performing a search with it. If you want to page over more than 10k objects, you are going to need to be able to open/close a PIT (without an API to do it implicitly for you).

So I guess this boils down to requirements:

1. Do we only want this feature to allow importing/exporting >10k SOs? In that case, we don't need to expose open/close PIT at all, it just means there is going to be an optional `pit` parameter on `SO.find` which is useless for folks outside of core.
2. Or do we want this feature to allow paging through >10k objects via `SO.find`? In that case, we need opening/closing PIT to have the same permissions as performing a SO.find against a particular type.

I've been working with the assumption that (2) is our end goal, but please chime in if anyone disagrees.

### Proposal
Continuing the assumption that our goal is (2), here is what I'd propose for next steps on this PR:
1. We remove `_pit` route from the server
1. We remove `openPointInTimeForType` from the browser APIs
1. We add `openPointInTimeForType` and `closePointInTime` to SO client on the server only
1. We put `openPointInTimeForType` and `closePointInTime` under the `read` permission group
1. We pass `namespace` down to both `openPointInTimeForType` and `closePointInTime` as @legrego previously suggested (even though they will be unused currently)

WDYT? Also - if anyone has suggestions on how to make this implicit in `SO.find` while still ensuring that it's possible to close a PIT, please chime in as that could simplify things 😄

Reply:
> I've been working with the assumption that (2) is our end goal, but please chime in if anyone disagrees.

+1, today plugins assume there will never be more than 10k results, and this will definitely break in ugly ways sooner or later. But no one should find and page through more than 10k results from the browser, plugins should use a different pattern like "only showing the first 10k results".

So I'm +1 for the proposal

The only use case this doesn't solve for is external users integrating with the HTTP API. It kinda sucks if you integrate with the API and then it stops working when you reach 10k objects. The users with the largest deployments are probably more likely to have API integrations. And then I guess the hesitation is, can a user easily de-stabilise a Kibana deployment accidentally and if so, there needs to be a way for admins to switch this off or ring fence the API. (We could start with a config option that's disabled by default).

Given that we can save users a lot of pain with the proposal as it is (because they can export 10k objects again) and all the uncertainty around exposing this as an HTTP API. I would say, let's leave the HTTP API out of scope and revisit it again if we hear that users have a need for it.

Reply:
> I've been working with the assumption that (2) is our end goal, but please chime in if anyone disagrees.

I'm +1 too.

The one thing I don't love is the ergonomics of starting a PIT. We have to call `openPointInTimeForType` with the same types that we expect the user to call `find` with. This has the potential to be error prone, but in practice it's likely not an issue; we only care about `type` for the PIT because we need to know which index to hit. So all that said, I think it's an acceptable tradeoff.

Reply:
+1 for me too

> I would say, let's leave the HTTP API out of scope and revisit it again if we hear that users have a need for it.

++

> We have to call openPointInTimeForType with the same types that we expect the user to call find with. This has the potential to be error prone, but in practice it's likely not an issue

Yea, in practice, you really have to break that in purpose imho.

Reply:
> We have to call openPointInTimeForType with the same types that we expect the user to call find with. This has the potential to be error prone, but in practice it's likely not an issue

This is true, but one other nuance to add on this point is that when you search using a PIT that's already been opened you aren't _allowed_ to specify an index or ES will 400. So in our case, we can either transparently drop any `types` that are provided to `SO.find` (current implementation), or we can explicitly throw an error if a user does this.

In the case of dropping `types` behind the scenes, it makes this somewhat less error prone: you just have to get the types right in your first call to open the PIT, and the rest is handled for you when you pass that PIT ID to `find`.

I'm moving forward with the proposal above since it sounds like we have some consensus on the approach here. I already pushed up an initial pass at updating the exporter to work with these new APIs if anyone wants to see how this can work in practice (still missing tests though).

Will ping when ready for a final review